### PR TITLE
ICU-12760 support UDISPCTX_TYPE_SUBSTITUTE_HANDLING

### PR DESCRIPTION
**Pull request by FrankYFTang**
##### Checklist

- [X] Issue filed: https://unicode-org.atlassian.net/browse/ICU-12760
- [X] Updated PR title and link in previous line to include Issue number
- [X] Issue accepted
- [X] Tests included
- [X] Documentation is changed or added

### CODE REVIEW

File: icu4c/source/common/locdspnm.cpp [hunk 1]
```diff
@@ -560,6 +572,7 @@ LocaleDisplayNamesImpl::adjustForUsageAndContext(CapContextUsage usage,
 UnicodeString&
 LocaleDisplayNamesImpl::localeDisplayName(const Locale& loc,
                                           UnicodeString& result) const {
+  printf("X\n");
```
Comment (fabalbon) [inline icu4c/source/common/locdspnm.cpp]: Please remove.

File: icu4c/source/test/intltest/locnmtst.cpp [hunk 2]
```diff
@@ -422,4 +423,59 @@ void LocaleDisplayNamesTest::TestRootEtc() {
...
+  test_assert_equal("wxyz", temp);
+  ldn->variantDisplayName("abc", temp);
+  test_assert_equal("abc", temp);
+  ldn->keyDisplayName("efg", temp);
+  test_assert_equal("efg", temp);
+  ldn->keyValueDisplayName("ca", "ijk", temp);
+  test_assert_equal("ijk", temp);
+}
+
+void LocaleDisplayNamesTest::VerifyNoSubstitute(LocaleDisplayNames* ldn) {
```
Comment (sffc) [inline icu4c/source/test/intltest/locnmtst.cpp]: Please add a test for some key that is present in root but not in the language bundle.

Comment (FrankYFTang) [reply icu4c/source/test/intltest/locnmtst.cpp]: I believe no such thing exist
see https://github.com/unicode-org/icu/blob/master/icu4c/source/data/lang/root.txt
and
https://github.com/unicode-org/icu/blob/master/icu4c/source/data/region/root.txt

Comment (sffc) [reply icu4c/source/test/intltest/locnmtst.cpp]: OK.

**Pull request merged**